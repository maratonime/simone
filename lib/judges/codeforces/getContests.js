'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getContests = getContests;

var _request = require('request');

var _request2 = _interopRequireDefault(_request);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getContests() {
  return new Promise(function (resolve, reject) {
    // making request to the codeforces api
    (0, _request2.default)('http://codeforces.com/api/contest.list', function (error, response, body) {
      if (response && response.statusCode == 200) {
        var contests = [];

        var data = JSON.parse(body).result;
        for (var index in data) {
          var info = data[index];

          // setting required fields
          var contest = {
            name: info.name,
            code: info.id,
            judge: 'codeforces',
            duration: info.durationSeconds,
            url: 'http://codeforces.com/contests/' + info.id
          };

          // setting fields that may be absent
          if ('startTimeSeconds' in info) contest.startTime = new Date(parseInt(info.startTimeSeconds) * 1000);
          if ('description' in info) contest.description = info.description;

          // setting contest state
          if (info.phase === 'BEFORE') contest.state = 'UPCOMING';else if (['CODING', 'PENDING_SYSTEM_TEST', 'SYSTEM_TEST'].includes(info.phase)) contest.state = 'RUNNING';else if (info.phase === 'FINISHED') contest.state = 'FINISHED';

          contests.push(contest);
        }
        resolve(contests);
      } else reject(error);
    });
  });
}