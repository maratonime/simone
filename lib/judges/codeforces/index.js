'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getContests = require('./getContests');

exports.default = {
  getContests: _getContests.getContests
};