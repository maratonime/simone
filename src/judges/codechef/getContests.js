// @flow
import { JSDOM } from 'jsdom';
import type { Contest } from '../../types';
import type { State } from '../../types';

function parse(row: HTMLTableRowElement): Contest {
  // flow has an issue concerning the cells field.
  // See https://github.com/facebook/flow/issues/6264
  const cells: HTMLCollection<HTMLTableCellElement> = (row: any).cells;

  const code: string = cells[0].textContent;
  const name: string = cells[1].children[0].textContent;
  const url: string = 'https://www.codechef.com/' + code;
  const startTime: Date = new Date(cells[2].dataset['starttime']);
  const endTime: Date = new Date(cells[3].dataset['endtime']);
  const duration: number = (endTime.getTime() - startTime.getTime()) / 1000;

  let state: State;
  const now = Date.now();
  if (startTime > now) state = 'UPCOMING';
  else if (endTime <= now) state = 'FINISHED';
  else state = 'RUNNING';

  return {
    name: name,
    code: code,
    judge: 'codechef',
    startTime: startTime,
    duration: duration,
    state: state,
    url: url,
  };
}

export async function getContests(): Promise<Array<Contest>> {
  // dom is a JSDOM, which is a third-party library that was not built with Flow
  const dom: any = await JSDOM.fromURL('https://www.codechef.com/contests');
  const document: Document = dom.window.document;

  const contestsTables: NodeList<HTMLElement> = document.querySelectorAll(
    'table.dataTable',
  );

  const contests: Array<Contest> = [];

  contestsTables.forEach(table => {
    const tab: HTMLTableElement = (table: any);
    const rows: HTMLCollection<HTMLTableRowElement> = tab.tBodies[0].rows;
    Array.from(rows).forEach(row => {
      contests.push(parse(row));
    });
  });

  return contests;
}
