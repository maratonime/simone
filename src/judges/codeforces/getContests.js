// @flow
import request from 'request-promise-native';
import type { Contest } from '../../types';
import { filterStartDate } from '../../utils';

type CFContest = {
  name: string,
  id: string,
  startTimeSeconds?: number,
  durationSeconds?: number,
  description?: string,
  phase: string,
};

export async function getContests(
  from?: Date,
  to?: Date,
): Promise<Array<Contest>> {
  const cfData: { result: Array<CFContest> } = JSON.parse(
    await request('http://codeforces.com/api/contest.list'),
  );

  return filterStartDate(
    cfData.result.map((contestData: CFContest) => {
      // setting required fields
      const contest: Contest = {
        name: contestData.name,
        code: contestData.id,
        judge: 'codeforces',
        url: 'http://codeforces.com/contests/' + contestData.id,
      };

      // setting fields that may be absent
      if (contestData.startTimeSeconds != null)
        contest.startTime = new Date(contestData.startTimeSeconds * 1000);
      if (contestData.durationSeconds != null)
        contest.duration = contestData.durationSeconds;
      if (contestData.description != null)
        contest.description = contestData.description;

      // setting contest state
      if (contestData.phase === 'BEFORE') contest.state = 'UPCOMING';
      else if (
        ['CODING', 'PENDING_SYSTEM_TEST', 'SYSTEM_TEST'].includes(
          contestData.phase,
        )
      )
        contest.state = 'RUNNING';
      else if (contestData.phase === 'FINISHED') contest.state = 'FINISHED';

      return contest;
    }),
    from,
    to,
  );
}
