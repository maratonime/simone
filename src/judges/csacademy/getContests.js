// @flow
import request from 'request-promise-native';
import type { Contest } from '../../types';

type CSContest = {|
  longName: string,
  name: string,
  id: string,
  startTime: ?string,
  endTime: ?string,
|};

export async function getContests(): Promise<Array<Contest>> {
  var options = {
    url: 'http://csacademy.com/contests/',
    headers: {
      'x-requested-with': 'XMLHttpRequest',
    },
  };
  const body = await request(options);
  const contests: Array<Contest> = [];

  const data: Object = JSON.parse(body).state.Contest;
  for (const info: CSContest of data) {
    if (
      info.longName === 'Archive' ||
      info.longName.startsWith('Virtual') || // Virtual contest for contestId=40468
      info.longName.startsWith('Contest') || // Contest for eval task 190 (Tree From Leaves)
      info.longName.match(/^\d\d:\d\d/) // 14:00, 28 September 2018 - algorithms
    )
      continue;
    // setting required fields
    const contest: Contest = {
      name: 'CSAcademy ' + info.longName,
      code: info.id,
      judge: 'csacademy',
      url: 'http://csacademy.com/contest/' + info.name,
    };

    // setting fields that may be absent
    if ('startTime' in info && info.startTime != null)
      contest.startTime = new Date(parseInt(info.startTime) * 1000);
    if (contest.startTime != null) {
      const startTime: Date = contest.startTime;
      const currentTime = Date.now();
      if ('endTime' in info && info.endTime != null) {
        const endTime: Date = new Date(parseInt(info.endTime) * 1000);
        contest.duration = Math.max(endTime - startTime, 0);
      }

      contest.state = currentTime < startTime ? 'UPCOMING' : 'RUNNING';
      if (
        contest.duration != null &&
        currentTime > new Date(startTime.getTime() + contest.duration)
      )
        contest.state = 'FINISHED';
    }

    contests.push(contest);
  }
  return contests;
}
