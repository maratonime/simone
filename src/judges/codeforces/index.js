// @flow
import { getContests } from './getContests';

export default {
  getContests: getContests
}