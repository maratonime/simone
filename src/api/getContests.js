// @flow

import type { Contest } from '../types/contest';
import type { GetContestsOptions } from '../types/api';
import judges from '../judges';

const default_options = Object.freeze({});

export async function getContests(
  options?: GetContestsOptions = default_options,
): Promise<Array<Contest>> {
  const contests: Array<Contest> = [];
  const promises: Array<Promise<Array<Contest>>> = [];
  Object.keys(judges).forEach(judge => {
    promises.push(
      judges[judge].getContests(options).catch(() => {
        console.error(judge + ' failed to getContests.');
        return [];
      }),
    );
  });
  const results = await Promise.all(promises);
  for (const contestList of results) contests.push(...contestList);
  return contests;
}
