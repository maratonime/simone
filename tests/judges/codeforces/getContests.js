jest.mock('request');
const request = require('request');
import { getContests } from './../../../src/judges/codeforces/getContests';

function mapToAPIReturn(...contests) {
  return JSON.stringify({
    status: 'OK',
    result: contests,
  });
}

function testContests(...contests) {
  request.mockReturnValueOnce(Promise.resolve(mapToAPIReturn(...contests)));
  return expect(getContests()).resolves.toMatchSnapshot();
}

test('simple contest', () => {
  const contest = {
    name: 'Oi',
    id: '123',
    durationSeconds: 321,
    phase: 'BEFORE',
  };
  return testContests(contest);
});

test('multiple contests', () => {
  const cs = [];
  for (let i = 1; i <= 10; i++)
    cs.push({
      name: 'Contest' + i,
      id: '' + i,
      durationSeconds: i * 60 * 60,
      phase: 'BEFORE',
    });
  return testContests(...cs);
});

test('filtering', () => {
  const baseContest = {
    name: '1',
    id: '123',
    durationSeconds: 111,
    phase: 'BEFORE',
  };
  const inside = [
    { ...baseContest, startTimeSeconds: 232 },
    { ...baseContest, startTimeSeconds: 234 },
  ];
  const outside = [
    { ...baseContest, startTimeSeconds: 230 },
    { ...baseContest, startTimeSeconds: 250 },
  ];
  request.mockReturnValueOnce(
    Promise.resolve(mapToAPIReturn(...inside, ...outside)),
  );
  return expect(
    getContests(new Date(231000), new Date(235000)),
  ).resolves.toMatchSnapshot();
});
