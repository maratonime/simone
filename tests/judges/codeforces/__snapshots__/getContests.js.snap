// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filtering 1`] = `
Array [
  Object {
    "code": "123",
    "duration": 111,
    "judge": "codeforces",
    "name": "1",
    "startTime": 1970-01-01T00:03:52.000Z,
    "state": "UPCOMING",
    "url": "http://codeforces.com/contests/123",
  },
  Object {
    "code": "123",
    "duration": 111,
    "judge": "codeforces",
    "name": "1",
    "startTime": 1970-01-01T00:03:54.000Z,
    "state": "UPCOMING",
    "url": "http://codeforces.com/contests/123",
  },
]
`;

exports[`multiple contests 1`] = `
Array [
  Object {
    "code": "1",
    "duration": 3600,
    "judge": "codeforces",
    "name": "Contest1",
    "state": "UPCOMING",
    "url": "http://codeforces.com/contests/1",
  },
  Object {
    "code": "2",
    "duration": 7200,
    "judge": "codeforces",
    "name": "Contest2",
    "state": "UPCOMING",
    "url": "http://codeforces.com/contests/2",
  },
  Object {
    "code": "3",
    "duration": 10800,
    "judge": "codeforces",
    "name": "Contest3",
    "state": "UPCOMING",
    "url": "http://codeforces.com/contests/3",
  },
  Object {
    "code": "4",
    "duration": 14400,
    "judge": "codeforces",
    "name": "Contest4",
    "state": "UPCOMING",
    "url": "http://codeforces.com/contests/4",
  },
  Object {
    "code": "5",
    "duration": 18000,
    "judge": "codeforces",
    "name": "Contest5",
    "state": "UPCOMING",
    "url": "http://codeforces.com/contests/5",
  },
  Object {
    "code": "6",
    "duration": 21600,
    "judge": "codeforces",
    "name": "Contest6",
    "state": "UPCOMING",
    "url": "http://codeforces.com/contests/6",
  },
  Object {
    "code": "7",
    "duration": 25200,
    "judge": "codeforces",
    "name": "Contest7",
    "state": "UPCOMING",
    "url": "http://codeforces.com/contests/7",
  },
  Object {
    "code": "8",
    "duration": 28800,
    "judge": "codeforces",
    "name": "Contest8",
    "state": "UPCOMING",
    "url": "http://codeforces.com/contests/8",
  },
  Object {
    "code": "9",
    "duration": 32400,
    "judge": "codeforces",
    "name": "Contest9",
    "state": "UPCOMING",
    "url": "http://codeforces.com/contests/9",
  },
  Object {
    "code": "10",
    "duration": 36000,
    "judge": "codeforces",
    "name": "Contest10",
    "state": "UPCOMING",
    "url": "http://codeforces.com/contests/10",
  },
]
`;

exports[`simple contest 1`] = `
Array [
  Object {
    "code": "123",
    "duration": 321,
    "judge": "codeforces",
    "name": "Oi",
    "state": "UPCOMING",
    "url": "http://codeforces.com/contests/123",
  },
]
`;
